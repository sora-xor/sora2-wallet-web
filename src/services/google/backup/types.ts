export enum BackupAccountType {
  PASSHRASE = 'PASSHRASE',
  JSON = 'JSON',
  SEED = 'SEED',
}

export type Json = {
  substrateJson?: string;
  ethJson?: string;
};

export type Seed = {
  substrateSeed?: string;
  ethSeed?: string;
};

export type DecryptedBackupAccount = {
  name: string;
  address: string;
  cryptoType: string;
  backupAccountType: BackupAccountType[];
  mnemonicPhrase: Nullable<string>;
  substrateDerivationPath: Nullable<string>;
  ethDerivationPath: Nullable<string>;
  seed: Nullable<Seed>;
  json: Nullable<Json>;
};

export type EncryptedBackupAccount = {
  name: string;
  address: string;
  cryptoType: string;
  backupAccountType: BackupAccountType[];
  encryptedMnemonicPhrase: Nullable<string>;
  encryptedEthDerivationPath: Nullable<string>;
  encryptedSubstrateDerivationPath: Nullable<string>;
  encryptedSeed: Nullable<Seed>;
  json: Nullable<Json>;
};
