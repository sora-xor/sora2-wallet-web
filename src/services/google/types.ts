export enum BackupAccountType {
  PASSHRASE = 'PASSHRASE',
  JSON = 'JSON',
  SEED = 'SEED',
}

export type Json = {
  substrateJson?: string;
  ethJson?: string;
};

export type Seed = {
  substrateSeed?: string;
  ethSeed?: string;
};

export type EncryptedBackupAccount = {
  name: string;
  address: string;
  encryptedMnemonicPhrase?: string;
  encryptedEthDerivationPath?: string;
  encryptedSubstrateDerivationPath?: string;
  cryptoType: string;
  backupAccountType: BackupAccountType[];
  encryptedSeed?: Seed;
  json?: Json;
};
