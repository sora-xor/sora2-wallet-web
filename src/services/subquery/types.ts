import type { History, CodecString, FPNumber } from '@sora-substrate/util';

import type { SoraNetwork } from '../../consts';

export interface ExplorerDataParser {
  parseTransactionAsHistoryItem: (transaction: any) => Promise<Nullable<History>>;
}

export interface Explorer {
  soraNetwork: SoraNetwork;
  getAccountTransactions: (variables?: any) => Promise<any>;
}

export enum ModuleNames {
  Assets = 'assets',
  LiquidityProxy = 'liquidityProxy',
  PoolXYK = 'poolXYK',
  TradingPair = 'tradingPair',
  Utility = 'utility',
  Referrals = 'referrals',
  EthBridge = 'ethBridge',
  BridgeMultisig = 'bridgeMultisig',
  Rewards = 'rewards',
  VestedRewards = 'vestedRewards',
  PswapDistribution = 'pswapDistribution',
}

export enum ModuleMethods {
  AssetsRegister = 'register',
  AssetsTransfer = 'transfer',
  PoolXYKInitializePool = 'initializePool',
  PoolXYKDepositLiquidity = 'depositLiquidity',
  PoolXYKWithdrawLiquidity = 'withdrawLiquidity',
  LiquidityProxySwap = 'swap',
  LiquidityProxySwapTransfer = 'swapTransfer',
  UtilityBatchAll = 'batchAll',
  ReferralsSetReferrer = 'setReferrer',
  ReferralsReserve = 'reserve',
  ReferralsUnreserve = 'unreserve',
  EthBridgeTransferToSidechain = 'transferToSidechain',
  BridgeMultisigAsMulti = 'asMulti',
  RewardsClaim = 'claim',
  VestedRewardsClaimRewards = 'claimRewards',
  VestedRewardsClaimCrowdloanRewards = 'claimCrowdloanRewards',
  PswapDistributionClaimIncentive = 'claimIncentive',
}

export type PoolXYKEntity = {
  strategicBonusApy: Nullable<string>;
  priceUSD: Nullable<string>;
  targetAssetId: string;
};

export type FiatPriceAndApyObject = {
  [key: string]: {
    price?: CodecString;
    strategicBonusApy?: CodecString;
  };
};

/**
 * `key` is timestamp, `value` is price USD
 */
export type HistoricalPrice = {
  [key: number]: CodecString;
};

export type HistoryElementError = {
  moduleErrorId: number;
  moduleErrorIndex: number;
};

export type HistoryElementExecution = {
  success: boolean;
  error?: HistoryElementError;
};

export interface HistoryElementSwap {
  baseAssetAmount: string;
  baseAssetId: string;
  liquidityProviderFee: string;
  selectedMarket: string;
  targetAssetAmount: string;
  targetAssetId: string;
}

export interface HistoryElementSwapTransfer extends HistoryElementSwap {
  to: string;
}

export type HistoryElementTransfer = {
  amount: string;
  assetId: string;
  from: string;
  to: string;
};

export type HistoryElementLiquidityOperation = {
  baseAssetAmount: string;
  baseAssetId: string;
  targetAssetAmount: string;
  targetAssetId: string;
  type: string;
};

export type HistoryElementAssetRegistration = {
  assetId: string;
};

export type ClaimedRewardItem = {
  assetId: string;
  amount: string;
};

export type HistoryElementRewardsClaim = Nullable<ClaimedRewardItem[]>;

export type UtilityBatchCall = {
  data: {
    args: {
      [key: string]: string | number;
    };
    callIndex: string;
    rewards?: ClaimedRewardItem[];
  };
  hash: string;
  callId: string;
  module: string;
  method: string;
};

export type ExtrinsicEvent = {
  method: string;
  section: string;
  data: any[];
};

export type HistoryElementUtilityBatchAll = UtilityBatchCall[];

export type HistoryElementEthBridgeOutgoing = {
  amount: string;
  assetId: string;
  sidechainAddress: string;
  requestHash?: string;
};

export type HistoryElementEthBridgeIncoming = {
  amount: string;
  assetId: string;
  requestHash: string;
  to: string;
};

export type HistoryElement = {
  id: string;
  blockHash: string;
  blockHeight: string;
  module: string;
  method: string;
  address: string;
  networkFee: string;
  execution: HistoryElementExecution;
  timestamp: number;
  data: Nullable<
    | HistoryElementSwap
    | HistoryElementSwapTransfer
    | HistoryElementTransfer
    | HistoryElementLiquidityOperation
    | HistoryElementAssetRegistration
    | HistoryElementUtilityBatchAll
    | ReferralSetReferrer
    | ReferrerReserve
    | HistoryElementEthBridgeOutgoing
    | HistoryElementEthBridgeIncoming
    | HistoryElementRewardsClaim
  >;
};

export type ReferrerRewards = {
  rewards: FPNumber;
  invitedUserRewards: {
    [key: string]: {
      rewards: FPNumber;
    };
  };
};

export type ReferrerReward = {
  id: string;
  blockHeight: string;
  referrer: string;
  referral: string;
  timestamp: number;
  amount: string;
};

export type ReferralSetReferrer = {
  from: string; // referral
  to: string; // referrer
};

export type ReferrerReserve = {
  from: string;
  to: string;
  amount: string;
};

export type ReferralRewardsGroup = {
  keys: [string];
  sum: {
    amount: CodecString;
  };
};
