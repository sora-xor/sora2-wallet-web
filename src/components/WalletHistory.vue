<template>
  <div class="history s-flex">
    <s-form
      class="history-form"
      :show-message="false"
    >
      <s-form-item v-if="hasHistory" class="history--search">
        <s-input
          v-model="query"
          :placeholder="t('history.filterPlaceholder')"
          prefix="el-icon-search"
          size="big"
          border-radius="mini"
        >
          <template #suffix v-if="query">
            <s-button class="s-button--clear" :use-design-system="false" @click="handleResetSearch">
              <s-icon name="clear-X-16" />
            </s-button>
          </template>
        </s-input>
      </s-form-item>
      <div class="history-items">
        <template v-if="filteredHistory.length">
          <div
            class="history-item s-flex"
            v-for="item in filteredHistory.slice((currentPage - 1) * pageAmount, currentPage * pageAmount)"
            :key="`history-${item.id}`"
            @click="handleOpenTransactionDetails(item.id)"
          >
            <div class="history-item-info">
              <div class="history-item-operation ch3" :data-type="item.type">{{ t(`operations.${item.type}`) }}</div>
              <div class="history-item-title p4">{{ getMessage(item) }}</div>
              <s-icon v-if="item.status !== TransactionStatus.Finalized" :class="getStatusClass(item.status)" :name="getStatusIcon(item.status)" />
            </div>
            <div class="history-item-date">{{ formatDate(item.startTime) }}</div>
          </div>
        </template>
        <div v-else class="history-empty p4">{{ t(`history.${hasHistory ? 'emptySearch' : 'empty'}`) }}</div>
      </div>
      <s-pagination
        v-if="hasHistory"
        :layout="'total, prev, next'"
        :current-page.sync="currentPage"
        :page-size="pageAmount"
        :total="filteredHistory.length"
        @prev-click="handlePaginationClick"
        @next-click="handlePaginationClick"
      />
    </s-form>
  </div>
</template>

<script lang="ts">
import { Component, Mixins, Prop } from 'vue-property-decorator'
import { Getter, Action } from 'vuex-class'

import { AccountAsset, History, TransactionStatus } from '@sora-substrate/util'
import LoadingMixin from './mixins/LoadingMixin'
import TransactionMixin from './mixins/TransactionMixin'
import { formatDate, getStatusIcon, getStatusClass } from '../util'
import { RouteNames } from '../consts'

@Component
export default class WalletHistory extends Mixins(LoadingMixin, TransactionMixin) {
  @Getter activity!: Array<History>
  @Action navigate
  @Action getAccountActivity

  @Prop() readonly asset?: AccountAsset

  formatDate = formatDate
  TransactionStatus = TransactionStatus
  query = ''
  currentPage = 1
  pageAmount = 8

  get assetAddress (): string | undefined {
    return this.asset && this.asset.address
  }

  get transactions (): Array<History> {
    if (this.assetAddress) {
      return this.activity.filter(item => [item.assetAddress, item.asset2Address].includes(this.assetAddress))
    }
    return this.activity
  }

  get filteredHistory (): Array<any> {
    if (!this.hasHistory) return []
    return this.getFilteredHistory(this.transactions).sort((a: History, b: History) => a.startTime && b.startTime ? b.startTime - a.startTime : 0)
  }

  get hasHistory (): boolean {
    return !!(this.transactions && this.transactions.length)
  }

  mounted () {
    this.getAccountActivity()
  }

  getFilteredHistory (history: Array<History>): Array<History> {
    if (!this.query) {
      return history
    }
    const query = this.query.toLowerCase().trim()
    return history.filter(item =>
      `${item.assetAddress}`.toLowerCase().includes(query) ||
      `${item.asset2Address}`.toLowerCase().includes(query) ||
      `${item.symbol}`.toLowerCase().includes(query) ||
      `${item.symbol2}`.toLowerCase().includes(query) ||
      `${item.blockId}`.toLowerCase().includes(query) ||
      `${item.from}`.toLowerCase().includes(query) ||
      `${item.to}`.toLowerCase().includes(query) ||
      this.t(`operations.${item.type}`).toLowerCase().includes(query)
    )
  }

  getStatus (status: string): string {
    if ([TransactionStatus.Error, 'invalid'].includes(status)) {
      status = TransactionStatus.Error
    } else if (status !== TransactionStatus.Finalized) {
      status = 'in_progress'
    }
    return status.toUpperCase()
  }

  getStatusClass (status: string): string {
    return getStatusClass(this.getStatus(status))
  }

  getStatusIcon (status: string): string {
    return getStatusIcon(this.getStatus(status))
  }

  handlePaginationClick (current: number): void {
    this.currentPage = current
  }

  handleOpenTransactionDetails (id: number): void {
    this.navigate({ name: RouteNames.WalletTransactionDetails, params: { id, asset: this.asset } })
  }

  handleResetSearch (): void {
    this.query = ''
    this.currentPage = 1
  }
}
</script>

<style lang="scss">
.history {
  margin-top: calc(var(--s-basic-spacing) * 2);
  .el-card__body {
    padding: calc(var(--s-basic-spacing) * 2) calc(var(--s-basic-spacing) * 2) calc(var(--s-basic-spacing) * 2.5);
  }
  .el-pagination {
    .btn {
      &-prev,
      &-next {
        padding-right: 0;
        padding-left: 0;
        min-width: calc(var(--s-basic-spacing) * 2.5);
      }
      &-prev {
        margin-left: auto;
        margin-right: var(--s-basic-spacing);
      }
    }
  }
  &--search {
    .el-input__inner {
      padding-right: var(--s-size-medium);
    }
  }
}
</style>

<style scoped lang="scss">
@import '../styles/icons';

$history-item-horizontal-space: 10px;
$history-item-height: 48px;
$history-item-top-border-height: 1px;

.history {
  flex-direction: column;
  &--search.el-form-item {
    margin-bottom: calc(var(--s-basic-spacing) * 2);
  }
  &-item {
    display: flex;
    flex-direction: column;
    margin-right: -#{$history-item-horizontal-space * 2};
    margin-left: -#{$history-item-horizontal-space * 2};
    min-height: $history-item-height;
    padding: calc(var(--s-basic-spacing) + #{$history-item-top-border-height}) $history-item-horizontal-space * 2;
    font-size: var(--s-font-size-mini);
    border-radius: var(--s-border-radius-small);
    &:not(:first-child) {
      position: relative;
      &:before {
        position: absolute;
        top: 0;
        left: 0;
        right: 0;
        display: block;
        margin-right: auto;
        margin-left: auto;
        height: 1px;
        width: calc(100% - #{$history-item-horizontal-space * 4});
        content: '';
        background-color: var(--s-color-base-border-secondary);
      }
    }
    &:hover {
      background-color: var(--s-color-base-background-hover);
      cursor: pointer;
      &:before, & + .history-item::before {
        width: 100%;
      }
    }
    &-info {
      display: flex;
      align-items: flex-start;
      .info-status--error  {
        color: var(--s-color-status-error);
      }
    }
    &-title {
      width: auto;
      padding-right: var(--s-basic-spacing);
      line-height: var(--s-line-height-mini);
    }
    &-title,
    &-date {
      width: 100%;
    }
    &-date {
      margin-top: calc(var(--s-basic-spacing) / 2);
      line-height: var(--s-line-height-mini);
      color: var(--s-color-base-content-tetriary);
    }
    &-operation {
      flex-shrink: 0;
      color: var(--s-color-theme-accent);
      border-radius: var(--s-border-radius-mini);
      margin-right: calc(var(--s-basic-spacing) / 2);
    }
    &-title,
    &-date {
      width: 100%;
    }
    &-date {
      color: var(--s-color-base-content-secondary);
      line-height: var(--s-line-height-mini);
    }
    &-icon {
      flex-shrink: 0;
      align-self: flex-start;
      margin-top: calc(var(--s-basic-spacing) / 2);
      margin-right: var(--s-basic-spacing);
      margin-left: auto;
    }
    .info-status--loading {
      @include svg-icon;
      @include loading;
      background-image: $status-pending-svg;
      height: var(--s-font-size-mini);
      width: var(--s-font-size-mini);
    }
  }
  &-empty {
    text-align: center;
  }
  .history--search {
    position: relative;
    .s-button--clear {
      width: 18px;
      height: 18px;
      padding: 0;
      background-color: transparent;
      border-radius: 0;
      border: none;
    }
  }
}
.el-pagination {
  display: flex;
  margin-top: calc(var(--s-basic-spacing) * 2);
  padding-left: 0;
  padding-right: 0;
}
</style>
