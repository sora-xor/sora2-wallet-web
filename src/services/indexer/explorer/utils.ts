import { formatStringNumber } from '../../../util';

import type { SubqueryAssetEntity } from '../subquery/types';
import type { SubsquidAssetEntity } from '../subsquid/types';
import type { FiatPriceObject, UpdatesStream, PoolApyObject } from '../types';

export function parseAssetFiatPrice(entity: SubsquidAssetEntity | SubqueryAssetEntity): FiatPriceObject {
  const acc = {};
  const id = entity.id;
  const priceFPNumber = formatStringNumber(entity.priceUSD);
  const isPriceFinity = priceFPNumber.isFinity();
  if (isPriceFinity) {
    acc[id] = priceFPNumber.toCodecString();
  }
  return acc;
}

export function parsePriceStreamUpdate(entity: UpdatesStream): Nullable<FiatPriceObject> {
  if (!entity?.data) return null;

  const data = JSON.parse(entity.data);

  return Object.entries(data).reduce((acc, [id, price]) => {
    const priceFPNumber = formatStringNumber(price as string);
    const isPriceFinity = priceFPNumber.isFinity();
    if (isPriceFinity) {
      acc[id] = priceFPNumber.toCodecString();
    }
    return acc;
  }, {});
}

export function parseApyStreamUpdate(entity: UpdatesStream): PoolApyObject {
  const data = entity?.data ? JSON.parse(entity.data) : {};

  return Object.entries(data).reduce((acc, [id, apy]) => {
    const strategicBonusApyFPNumber = formatStringNumber(apy as string);
    const isStrategicBonusApyFinity = strategicBonusApyFPNumber.isFinity();
    if (isStrategicBonusApyFinity) {
      acc[id] = strategicBonusApyFPNumber.toCodecString();
    }
    return acc;
  }, {});
}
