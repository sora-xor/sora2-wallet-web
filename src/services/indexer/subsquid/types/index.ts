export * from '../../types';
export * from './models';
export * from './response';
export * from './subsquid';
