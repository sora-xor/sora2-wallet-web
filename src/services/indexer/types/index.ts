export * from './calls';
// export * from './models';
