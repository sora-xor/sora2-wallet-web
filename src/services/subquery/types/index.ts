export * from './app';
export * from './calls';
export * from './models';
export * from './response';
export * from './subquery';
