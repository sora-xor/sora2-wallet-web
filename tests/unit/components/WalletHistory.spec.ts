import Vuex from 'vuex';
import { shallowMount, mount } from '@vue/test-utils';
import { Account } from '@/types/common';
import { History, TransactionStatus } from '@sora-substrate/util';

import WalletHistory from '@/components/WalletHistory.vue';
import { useDescribe, localVue, i18n } from '../../utils';
import { MOCK_ACCOUNT_ASSETS, MOCK_HISTORY } from '../../utils/mock';
import { MOCK_ACCOUNT } from '../../utils/WalletAccountMock';

const createStore = (historyArray?: Array<History> | undefined) => {
  return new Vuex.Store({
    getters: {
      account: () => MOCK_ACCOUNT as Account,
      activity: () => historyArray as Array<History>,
      shouldBalanceBeHidden: () => true as boolean,
    },
    actions: {
      getAccountActivity: jest.fn(),
    },
  });
};

useDescribe('WalletHistory.vue', WalletHistory, () => {
  it('should be rendered correctly', () => {
    const wrapper = shallowMount(WalletHistory, {
      localVue,
      store: createStore(MOCK_HISTORY),
      i18n,
    });

    expect(wrapper.element).toMatchSnapshot();
  });

  it('should render history empty text when no transacion', () => {
    const wrapper = shallowMount(WalletHistory, { localVue, store: createStore(), i18n });
    const textMessage = wrapper.find('.history-empty');

    expect(textMessage.exists()).toBeTrue();
    expect(wrapper.element).toMatchSnapshot();
  });

  it('asset should be readable when passed as prop', () => {
    const wrapper = shallowMount(WalletHistory, {
      localVue,
      store: createStore(MOCK_HISTORY),
      i18n,
      propsData: { asset: MOCK_ACCOUNT_ASSETS[0] },
    });

    expect(wrapper.vm.$props.asset).toEqual(MOCK_ACCOUNT_ASSETS[0]);
  });

  it('should render spinner when transaction is pending', () => {
    const wrapper = mount(WalletHistory, {
      localVue,
      store: createStore([{ ...MOCK_HISTORY[0], status: TransactionStatus.InBlock }]),
      i18n,
      propsData: { asset: MOCK_ACCOUNT_ASSETS[0] },
    });
    const pendingIcon = wrapper.find('.info-status--loading');

    expect(pendingIcon.exists()).toBeTrue();
    expect(wrapper.element).toMatchSnapshot();
  });

  it('should render error icon when transaction is failed', () => {
    const wrapper = mount(WalletHistory, {
      localVue,
      store: createStore([{ ...MOCK_HISTORY[0], status: TransactionStatus.Error }]),
      i18n,
      propsData: { asset: MOCK_ACCOUNT_ASSETS[0] },
    });
    const errorIcon = wrapper.find('.info-status--error');

    expect(errorIcon.exists()).toBeTrue();
    expect(wrapper.element).toMatchSnapshot();
  });
});
