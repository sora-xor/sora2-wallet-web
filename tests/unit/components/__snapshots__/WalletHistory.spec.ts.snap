// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`WalletHistory.vue should be rendered correctly 1`] = `
<div
  class="history s-flex"
>
  <s-form-stub
    class="history-form"
    labelposition="right"
    validateonrulechange="true"
  >
    <search-input-stub
      autofocus="true"
      class="history--search"
      placeholder="Filter by Address, Symbol, Type or Block ID"
      value=""
    />
     
    <div
      class="history-items"
    >
      <div
        class="history-item s-flex"
        role="button"
        tabindex="0"
      >
        <div
          class="history-item-info"
        >
          <div
            class="history-item-operation ch3"
            data-type="Transfer"
          >
            Transfer
          </div>
           
          <div
            class="history-item-title p4"
          >
            Sent ****** VAL To cnW4q...L2ftG
          </div>
           
          <!---->
        </div>
         
        <div
          class="history-item-date"
        >
          Nov 27, 2020 11:59:59 PM
        </div>
      </div>
      <div
        class="history-item s-flex"
        role="button"
        tabindex="0"
      >
        <div
          class="history-item-info"
        >
          <div
            class="history-item-operation ch3"
            data-type="Transfer"
          >
            Transfer
          </div>
           
          <div
            class="history-item-title p4"
          >
            Sent ****** XOR To cnW4q...L2ftG
          </div>
           
          <!---->
        </div>
         
        <div
          class="history-item-date"
        >
          Nov 27, 2020 11:59:59 PM
        </div>
      </div>
      <div
        class="history-item s-flex"
        role="button"
        tabindex="0"
      >
        <div
          class="history-item-info"
        >
          <div
            class="history-item-operation ch3"
            data-type="Transfer"
          >
            Transfer
          </div>
           
          <div
            class="history-item-title p4"
          >
            Sent ****** XOR To cnW4q...L2ftG
          </div>
           
          <!---->
        </div>
         
        <div
          class="history-item-date"
        >
          Nov 27, 2020 11:59:59 PM
        </div>
      </div>
      <div
        class="history-item s-flex"
        role="button"
        tabindex="0"
      >
        <div
          class="history-item-info"
        >
          <div
            class="history-item-operation ch3"
            data-type="Transfer"
          >
            Transfer
          </div>
           
          <div
            class="history-item-title p4"
          >
            Sent ****** VAL To cnW4q...L2ftG
          </div>
           
          <!---->
        </div>
         
        <div
          class="history-item-date"
        >
          Nov 27, 2020 11:59:59 PM
        </div>
      </div>
      <div
        class="history-item s-flex"
        role="button"
        tabindex="0"
      >
        <div
          class="history-item-info"
        >
          <div
            class="history-item-operation ch3"
            data-type="Transfer"
          >
            Transfer
          </div>
           
          <div
            class="history-item-title p4"
          >
            Sent ****** XOR To cnVFi...P2nse
          </div>
           
          <!---->
        </div>
         
        <div
          class="history-item-date"
        >
          Nov 27, 2020 11:59:59 PM
        </div>
      </div>
      <div
        class="history-item s-flex"
        role="button"
        tabindex="0"
      >
        <div
          class="history-item-info"
        >
          <div
            class="history-item-operation ch3"
            data-type="Swap"
          >
            Swap
          </div>
           
          <div
            class="history-item-title p4"
          >
            Failed to swap ****** XOR for ****** PSWAP
          </div>
           
          <s-icon-stub
            class="info-status info-status--error"
            name="basic-clear-X-xs-24"
          />
        </div>
         
        <div
          class="history-item-date"
        >
          Nov 27, 2020 11:59:59 PM
        </div>
      </div>
      <div
        class="history-item s-flex"
        role="button"
        tabindex="0"
      >
        <div
          class="history-item-info"
        >
          <div
            class="history-item-operation ch3"
            data-type="AddLiquidity"
          >
            Add Liquidity
          </div>
           
          <div
            class="history-item-title p4"
          >
            Supplied ****** XOR and ****** PSWAP
          </div>
           
          <!---->
        </div>
         
        <div
          class="history-item-date"
        >
          Nov 27, 2020 11:59:59 PM
        </div>
      </div>
      <div
        class="history-item s-flex"
        role="button"
        tabindex="0"
      >
        <div
          class="history-item-info"
        >
          <div
            class="history-item-operation ch3"
            data-type="RemoveLiquidity"
          >
            Remove Liquidity
          </div>
           
          <div
            class="history-item-title p4"
          >
            Removed ****** XOR and ****** PSWAP
          </div>
           
          <!---->
        </div>
         
        <div
          class="history-item-date"
        >
          Nov 27, 2020 11:59:59 PM
        </div>
      </div>
    </div>
     
    <history-pagination-stub
      currentpage="1"
      lastpage="2"
      pageamount="8"
      total="9"
    />
  </s-form-stub>
</div>
`;

exports[`WalletHistory.vue should render error icon when transaction is failed 1`] = `
<div
  class="history s-flex"
>
  <s-form-stub
    class="history-form"
    labelposition="right"
    validateonrulechange="true"
  >
    <search-input-stub
      autofocus="true"
      class="history--search"
      placeholder="Filter by Address, Symbol, Type or Block ID"
      value=""
    />
     
    <div
      class="history-items"
    >
      <div
        class="history-item s-flex"
        role="button"
        tabindex="0"
      >
        <div
          class="history-item-info"
        >
          <div
            class="history-item-operation ch3"
            data-type="Transfer"
          >
            Transfer
          </div>
           
          <div
            class="history-item-title p4"
          >
            Sent ****** VAL To cnW4q...L2ftG
          </div>
           
          <!---->
        </div>
         
        <div
          class="history-item-date"
        >
          Nov 27, 2020 11:59:59 PM
        </div>
      </div>
      <div
        class="history-item s-flex"
        role="button"
        tabindex="0"
      >
        <div
          class="history-item-info"
        >
          <div
            class="history-item-operation ch3"
            data-type="Transfer"
          >
            Transfer
          </div>
           
          <div
            class="history-item-title p4"
          >
            Sent ****** XOR To cnW4q...L2ftG
          </div>
           
          <!---->
        </div>
         
        <div
          class="history-item-date"
        >
          Nov 27, 2020 11:59:59 PM
        </div>
      </div>
      <div
        class="history-item s-flex"
        role="button"
        tabindex="0"
      >
        <div
          class="history-item-info"
        >
          <div
            class="history-item-operation ch3"
            data-type="Transfer"
          >
            Transfer
          </div>
           
          <div
            class="history-item-title p4"
          >
            Sent ****** XOR To cnW4q...L2ftG
          </div>
           
          <!---->
        </div>
         
        <div
          class="history-item-date"
        >
          Nov 27, 2020 11:59:59 PM
        </div>
      </div>
      <div
        class="history-item s-flex"
        role="button"
        tabindex="0"
      >
        <div
          class="history-item-info"
        >
          <div
            class="history-item-operation ch3"
            data-type="Transfer"
          >
            Transfer
          </div>
           
          <div
            class="history-item-title p4"
          >
            Sent ****** VAL To cnW4q...L2ftG
          </div>
           
          <!---->
        </div>
         
        <div
          class="history-item-date"
        >
          Nov 27, 2020 11:59:59 PM
        </div>
      </div>
      <div
        class="history-item s-flex"
        role="button"
        tabindex="0"
      >
        <div
          class="history-item-info"
        >
          <div
            class="history-item-operation ch3"
            data-type="Transfer"
          >
            Transfer
          </div>
           
          <div
            class="history-item-title p4"
          >
            Failed to send ****** XOR to cnVFi...P2nse
          </div>
           
          <s-icon-stub
            class="info-status info-status--error"
            name="basic-clear-X-xs-24"
          />
        </div>
         
        <div
          class="history-item-date"
        >
          Nov 27, 2020 11:59:59 PM
        </div>
      </div>
    </div>
     
    <!---->
  </s-form-stub>
</div>
`;

exports[`WalletHistory.vue should render history empty text when no transacion 1`] = `
<div
  class="history s-flex"
>
  <s-form-stub
    class="history-form"
    labelposition="right"
    validateonrulechange="true"
  >
    <!---->
     
    <div
      class="history-items"
    >
      <div
        class="history-empty p4"
      >
        Your transactions will appear here
      </div>
    </div>
     
    <!---->
  </s-form-stub>
</div>
`;

exports[`WalletHistory.vue should render spinner when transaction is pending 1`] = `
<div
  class="history s-flex"
>
  <s-form-stub
    class="history-form"
    labelposition="right"
    validateonrulechange="true"
  >
    <search-input-stub
      autofocus="true"
      class="history--search"
      placeholder="Filter by Address, Symbol, Type or Block ID"
      value=""
    />
     
    <div
      class="history-items"
    >
      <div
        class="history-item s-flex"
        role="button"
        tabindex="0"
      >
        <div
          class="history-item-info"
        >
          <div
            class="history-item-operation ch3"
            data-type="Transfer"
          >
            Transfer
          </div>
           
          <div
            class="history-item-title p4"
          >
            Sent ****** VAL To cnW4q...L2ftG
          </div>
           
          <!---->
        </div>
         
        <div
          class="history-item-date"
        >
          Nov 27, 2020 11:59:59 PM
        </div>
      </div>
      <div
        class="history-item s-flex"
        role="button"
        tabindex="0"
      >
        <div
          class="history-item-info"
        >
          <div
            class="history-item-operation ch3"
            data-type="Transfer"
          >
            Transfer
          </div>
           
          <div
            class="history-item-title p4"
          >
            Sent ****** XOR To cnW4q...L2ftG
          </div>
           
          <!---->
        </div>
         
        <div
          class="history-item-date"
        >
          Nov 27, 2020 11:59:59 PM
        </div>
      </div>
      <div
        class="history-item s-flex"
        role="button"
        tabindex="0"
      >
        <div
          class="history-item-info"
        >
          <div
            class="history-item-operation ch3"
            data-type="Transfer"
          >
            Transfer
          </div>
           
          <div
            class="history-item-title p4"
          >
            Sent ****** XOR To cnW4q...L2ftG
          </div>
           
          <!---->
        </div>
         
        <div
          class="history-item-date"
        >
          Nov 27, 2020 11:59:59 PM
        </div>
      </div>
      <div
        class="history-item s-flex"
        role="button"
        tabindex="0"
      >
        <div
          class="history-item-info"
        >
          <div
            class="history-item-operation ch3"
            data-type="Transfer"
          >
            Transfer
          </div>
           
          <div
            class="history-item-title p4"
          >
            Sent ****** VAL To cnW4q...L2ftG
          </div>
           
          <!---->
        </div>
         
        <div
          class="history-item-date"
        >
          Nov 27, 2020 11:59:59 PM
        </div>
      </div>
      <div
        class="history-item s-flex"
        role="button"
        tabindex="0"
      >
        <div
          class="history-item-info"
        >
          <div
            class="history-item-operation ch3"
            data-type="Transfer"
          >
            Transfer
          </div>
           
          <div
            class="history-item-title p4"
          >
            Sent ****** XOR To cnVFi...P2nse
          </div>
           
          <s-icon-stub
            class="info-status info-status--loading"
            name="refresh-16"
          />
        </div>
         
        <div
          class="history-item-date"
        >
          Nov 27, 2020 11:59:59 PM
        </div>
      </div>
    </div>
     
    <!---->
  </s-form-stub>
</div>
`;
